### PR TITLE
Rework on dependency downloading

### PR DESCRIPTION
**Pull request by zfy0701**

### CODE REVIEW

File: src/dependency-manager.ts [hunk 1]
```diff
@@ -87,4 +104,103 @@ export class DependencyManager {
...
+            const lib = url.protocol.startsWith('https') ? https : http;
+            const request = lib.get(url, (response: http.IncomingMessage) => {
+                if (!response.statusCode || response.statusCode < 200 || response.statusCode > 299) {
+                    reject(new Error('Failed to load page, status code: ' + response.statusCode));
+                }
+
+                const extractorStream: internal.Writable = gunzip();
+                const untarStream = tarFs.extract(dest, {
+                    // @ts-ignore
+                    strip: 1,
```
Comment (alexbrasetvik) [inline src/dependency-manager.ts:161]: Probably want to make sure we don't extract symlinks here, and as always that paths are checked for `../`

Please look at these resources:

- https://snyk.io/blog/severe-security-vulnerability-in-bowers-zip-archive-extraction 
- https://snyk.io/research/zip-slip-vulnerability
- https://hackerone.com/reports/473811

**Review by alexbrasetvik — CHANGES_REQUESTED** [review]

File: src/dependency-manager.ts (hunk 1 above)
Comment (alexbrasetvik) [reply src/dependency-manager.ts:161]: I'm guessing tarFs will err out if the stream isn't a valid tar archive.

If not, what happens if http://169.254.169.254/latest/user-data is provided as a package URL?

Comment (zfy0701) [reply src/dependency-manager.ts:161]: done, and it's can't be an arbitrary file, it will just error out

File: src/dependency-manager.ts [hunk 2]
```diff
@@ -87,4 +104,112 @@ export class DependencyManager {
...
+            }
+        } catch (e) {
+            this.logger.error(e);
+        }
+    }
+
+    public async downloadFile(url: URL, dest: string): Promise<void> {
+        return new Promise((resolve, reject) => {
+            const lib = url.protocol.startsWith('https') ? https : http;
+            const request = lib.get(url, (response: http.IncomingMessage) => {
```
Comment (spacedragon) [inline src/dependency-manager.ts:144]: I think `https.get` won't handle HTTP redirects, you may need to handle status codes 301/302 yourself or use a library.

https://stackoverflow.com/questions/7323932/how-do-you-follow-an-http-redirect-in-node-js

**Review by spacedragon — APPROVED** [review]
Left a comment inline.

**Pull request closed without merging**